File: 00.My_Practice_Courses/03.Algorithms_Java_Solutions/04.Dynamic_Programming/DynamicProgrammingLAB/src/LCS/LongestCommonSubsequence.java
package LCS;

/**
 * Longest common subsequence (LCS) problem:
 * Given two sequences x[1 … m] and y[1 … n]
 * Find a longest common subsequence (LCS) to them both
 *
 * x = "ABCBDAB"
 * y = "BDCABA"
 * LCS = "BCBA"
 */
public class LongestCommonSubsequence {

    public static void main(String[] args) {
        String first = "tea";
        String second = "tree";

        int[][] lcsMatrix = findLcs(first, second);
        for (int row = 0; row < lcsMatrix.length; row++) {
            for (int col = 0; col < lcsMatrix[row].length;col++) {
                System.out.print(lcsMatrix[row][col] + " ");
            }
            System.out.println();
        }
        String lcs = retrieveLcs(first, second, lcsMatrix);

        System.out.println(String.format("First string: %s", first));
        System.out.println(String.format("Second string: %s", second));
        System.out.println(String.format("Longest common subsequence: %s", lcs));
        System.out.println(String.format("Length: %d", lcs.length()));
    }

    private static int[][] findLcs(String first, String second) {

        int[][] lcsMatrix = new int[first.length() + 1][second.length() + 1];

        for (int row = 1; row < lcsMatrix.length; row++) {
            for (int col = 1; col < lcsMatrix[row].length; col++) {

                if (first.charAt(row - 1) == second.charAt(col - 1)) {
                    int diagonalCell = lcsMatrix[row - 1][col - 1];
                    lcsMatrix[row][col] = diagonalCell + 1;
                } else {
                    int upCell = lcsMatrix[row - 1][col];
                    int leftCell = lcsMatrix[row][col - 1];
                    lcsMatrix[row][col] =
                            Math.max(upCell, leftCell);
                }
            }
        }

        return lcsMatrix;
    }

    private static String retrieveLcs(String first, String second, int[][] lcsMatrix) {
        StringBuilder lcs = new StringBuilder();
        int row = lcsMatrix.length - 1;
        int col = lcsMatrix[0].length - 1;

        while (row > 0 && col > 0) {
            if (first.charAt(row - 1) == second.charAt(col - 1)) {
                lcs.insert(0, first.charAt(row - 1));
                row--;
                col--;
            } else if (lcsMatrix[row][col] == lcsMatrix[row - 1][col]) {
                row--;
            } else {
                col--;
            }
        }

        return lcs.toString();
    }
}

File: 00.My_Practice_Courses/02Algorithms_April_2016/04.Dynamic_Programming_Lab/Longest-Common-Subsequence/LongestCommonSubsequence.cs
namespace Longest_Common_Subsequence
{
    using System;
    using System.Collections.Generic;

    public class LongestCommonSubsequence
    {
        public static void Main()
        {
            Console.Write("First string = ");
            var firstStr = Console.ReadLine();
            Console.Write("Second string = ");
            var secondStr = Console.ReadLine();

            var lcs = FindLongestCommonSubsequence(firstStr, secondStr);

            Console.WriteLine("Longest common subsequence:");
            Console.WriteLine("  first  = {0}", firstStr);
            Console.WriteLine("  second = {0}", secondStr);
            Console.WriteLine("  lcs    = {0}", lcs);
        }

        public static string FindLongestCommonSubsequence(string firstStr, string secondStr)
        {
            var firstLen = firstStr.Length + 1;
            var secondLen = secondStr.Length + 1;
            var lcs = new int[firstLen, secondLen];

            for (int row = 1; row < lcs.GetLength(0); row++)
            {
                for (int col = 1; col < lcs.GetLength(1); col++)
                {
                    if (firstStr[row - 1] == secondStr[col - 1])
                    {
                        lcs[row, col] = lcs[row - 1, col - 1] + 1;
                    }
                    else
                    {
                        lcs[row, col] = Math.Max(lcs[row - 1, col], lcs[row, col - 1]);
                    }
                }
            }
            string sequence = RetrieveLCS(firstStr, secondStr, lcs);
            return sequence;
        }

        private static string RetrieveLCS(string firstStr, string secondStr, int[,] lcs)
        {
            int row = firstStr.Length;
            int col = secondStr.Length;
            var sequence = new List<char>();

            while (row > 0 && col > 0)
            {
                if (firstStr[row - 1] == secondStr[col - 1])
                {
                    sequence.Add(firstStr[row - 1]);
                    row--;
                    col--;
                }
                else if (lcs[row, col] == lcs[row - 1, col])
                {
                    row--;
                }
                else
                {
                    col--;
                }
            }
            sequence.Reverse();
            return new string(sequence.ToArray());
        }
    }
}
